CXF-7742 handling scopes in CdiResourceProvider

rmannibucau opened the pull request.

reta commented on integration/cdi/src/main/java/org/apache/cxf/cdi/JAXRSCdiResourceExtension.java:
```diff
@@ -303,6 +306,10 @@ public void release(@Observes final BeforeShutdown event) {
+            disposableCreationalContexts.clear();
+        }
+        synchronized (disposableResourceProviders) {
+            disposableResourceProviders.forEach(p -> p.getLifecycle().destroy());
```
`disposableResourceProviders.clear()`?

reta commented on integration/cdi/src/main/java/org/apache/cxf/cdi/JAXRSCdiResourceExtension.java:
```diff
@@ -372,7 +379,14 @@ private ClassifiedClasses classes2singletons(final Application application, fina
+                    // if @Singleton the container handles it
+                    // so we only need this case here
+                    if (provider.isSingleton() && Dependent.class == bean.getScope()) {
+                        disposableResourceProviders.add(provider);
```
```
synchronized (disposableResourceProviders) {
    disposableResourceProviders.add(provider);
}
```

rmannibucau replied:
Oh good catch, copied the pattern of disposableCreationalContexts but actually we don't need that for that list which is just used in the context of the extension so thread safe, will clean up

reta commented:
@rmannibucau Looks good, a few minor comments, I am wondering if it worth creating a test case for different resource providers?

rmannibucau commented:
@reta I updated the PR, I split the provider for singletons and per request handling, it looks easier to me to follow what is done and avoid to be tempted to be back to something not thread safe, I added a test on the scope handling too

reta commented on integration/cdi/src/main/java/org/apache/cxf/cdi/SingletonResourceProvider.java:
```diff
@@ -0,0 +1,52 @@
+    private Object instance;
+
+    SingletonResourceProvider(final Lifecycle lifecycle, final Class<?> clazz) {
+        this.instance = lifecycle.create();
```
Should we create instance proactively here? It is simpler like that but it may not be needed (in theory, at least)

rmannibucau replied:
Well for all normal scoped beans it is still lazy thanks to CDI proxies and for @Singleton/@Dependent it becomes eager but it is ok from a business perspective + the extension already does it in several cases so not a big deal IMHO.

reta commented:
@rmannibucau Looks better like that, with providers split between different classes

rmannibucau commented:
Anything preventing a merge or are we good to go like that? (No pression, just checking if something is expected from my side)

reta commented:
@rmannibucau Looks good to me, do you want me to merge?

rmannibucau commented:
@reta if possible, sure. Thanks a lot.

reta commented:
Done, thanks a lot for fixing that, @rmannibucau !
